In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [14]:
data = pd.read_csv('../data/bank-additional-full.csv',sep=';')

In [21]:
data.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

In [22]:
conversion_rate = (data['y']=='yes').mean()*100
conversion_rate

np.float64(11.265417111780131)

In [26]:
data.groupby('contact')['y'].apply(lambda x : (x=='yes').mean()*100)

contact
cellular     14.737607
telephone     5.231321
Name: y, dtype: float64

In [30]:
data.groupby('y')['duration'].mean()

y
no     220.844807
yes    553.191164
Name: duration, dtype: float64

In [33]:
data.groupby('month')['y'].apply(lambda x : (x=='yes').mean()*100).sort_values(ascending=False)

month
mar    50.549451
dec    48.901099
sep    44.912281
oct    43.871866
apr    20.478723
aug    10.602137
jun    10.511470
nov    10.143867
jul     9.046557
may     6.434745
Name: y, dtype: float64

In [34]:
data['month'].value_counts()

month
may    13769
jul     7174
aug     6178
jun     5318
nov     4101
apr     2632
oct      718
sep      570
mar      546
dec      182
Name: count, dtype: int64

In [38]:
data.groupby('campaign')['y'].apply(lambda x : (x=='yes').mean()*100).head(10)

campaign
1     13.037071
2     11.456954
3     10.747051
4      9.392682
5      7.504690
6      7.660878
7      6.041335
8      4.250000
9      6.007067
10     5.333333
Name: y, dtype: float64

In [39]:
data.groupby('poutcome')['y'].apply(lambda x : (x=='yes').mean()*100)

poutcome
failure        14.228598
nonexistent     8.832213
success        65.112891
Name: y, dtype: float64

# 🔎 Marketing Funnel & Conversion Insights

## 1️⃣ Overall Conversion Performance
The overall campaign conversion rate is approximately **11.27%**, indicating that the majority of contacted customers did not subscribe. This suggests significant room for optimization in targeting and campaign strategy.

---

## 2️⃣ Contact Channel Effectiveness
Conversion rates vary significantly by contact method:

- 📱 Cellular: ~14.74%
- ☎️ Telephone: ~5.23%

Cellular contact performs nearly **3x better** than telephone, indicating that mobile outreach is significantly more effective and should be prioritized.

---

## 3️⃣ Call Engagement Impact
Average call duration differs substantially between converted and non-converted customers:

- Converted customers: ~553 seconds
- Non-converted customers: ~221 seconds

Longer call duration strongly correlates with successful conversions, highlighting the importance of sales engagement quality and conversation depth.

---

## 4️⃣ Campaign Frequency Effect
Conversion rates decline as the number of contact attempts increases:

- 1st contact: ~13%
- 5th contact: ~7.5%
- 8th contact: ~4%

This suggests diminishing returns from repeated outreach. Customers who convert typically do so within the first few calls.

---

## 5️⃣ Seasonal Performance Patterns
Certain months (e.g., March, December) show high conversion rates, but these months also have significantly lower contact volumes. High-volume months such as May show lower efficiency (~6.4%), indicating possible over-targeting or weaker campaign quality.

---

## 6️⃣ Previous Campaign Outcome (Retargeting Insight)
Conversion rates vary dramatically based on previous campaign results:

- Previous Success: **65.1%**
- Previous Failure: 14.2%
- No Previous Contact: 8.8%

Customers who previously responded positively are highly likely to convert again, indicating strong potential for retargeting and loyalty-focused strategies.

---

# 💡 Strategic Recommendations

## ✅ Prioritize Cellular Outreach
Allocate more marketing resources toward mobile-based campaigns to maximize conversion efficiency.

## ✅ Improve First-Call Quality
Since most conversions occur within the first few interactions, sales teams should focus on optimizing initial contact scripts and engagement techniques.

## ✅ Limit Excessive Repeated Contacts
Reduce repeated outreach beyond 3–4 attempts to avoid diminishing returns and reallocate effort toward new prospects.

## ✅ Implement Retargeting Strategy
Actively target customers with previous successful interactions, as they represent the highest-probability conversion segment.

## ✅ Optimize High-Volume Campaign Months
Review targeting strategy during peak months (e.g., May) to improve efficiency and avoid quality dilution.

---

# 🚀 Strategic Conclusion

This funnel analysis demonstrates that conversion performance is heavily influenced by contact method, engagement quality, campaign frequency, and prior customer interactions.

By focusing on high-quality targeting, retargeting successful leads, and optimizing early-stage engagement, the bank can significantly improve overall conversion rates and marketing ROI.